src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs

```diff
@@ -241,7 +241,38 @@ int IList.Add(object? item)
         // capacity or the new size, whichever is larger.
         //
         public void AddRange(IEnumerable<T> collection)
-            => InsertRange(_size, collection);
+        {
+            if (collection == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.collection);
+            }
+
+            if (collection is ICollection<T> c)
+            {
+                int count = c.Count;
+                if (count > 0)
+                {
+                    if (_items.Length - _size < count)
+                    {
+                        Grow(_size + count);
+                    }
+
+                    c.CopyTo(_items, _size);
```

Comment:
I'm guessing this could regress cases where we're appending the list to itself?

Reply:
Regress functionally? We have this test:
https://github.com/dotnet/runtime/blob/9b9b0c95c678065044d01049c56915e6b750b10a/src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.AddRange.cs#L50
Is there a case that's missing?

Reply:
Functionality looks good, I meant to say regress performance.

Reply:
I don't believe so.  I just tried with:
```C#
[Benchmark]
public List<int> AddToSelf()
{
    var list = new List<int>();
    list.Add(1);
    list.Add(2);
    list.AddRange(list);
    list.AddRange(list);
    list.AddRange(list);
    list.AddRange(list);
    list.AddRange(list);
    list.AddRange(list);
    return list;
}
```
and depending on the run there was either no difference or a small difference in favor in of the new version.

Is there a specific aspect of the change you think would regress that? I can try to address it if so. Though I'm not particularly concerned about the performance of that case, anyway.

Reply:
Assuming we cared about optimizing that particular case, hardcoding the operation in the style of `InsertRange` might be marginally faster compared to using `ICollection<T>.CopyTo`, at the expense of added branching for every other case. But like you said, it doesn't seem important enough to justify.